design/namespaced-roles.md

```diff
@@ -0,0 +1,299 @@
+
+### Use Case 1 - Assign a role in the context of a group
+
+Say we want to have the following group composition in our realm (or our tenant, in the future), most likely imported from a federated LDAP:
+
+realm/tenant:tenant1
+ - group:iam
+ - group:devops
+ - group:leadership
+
+Following the current implementation, the users would need to define roles at the realm level and then assign them to these groups. Everyone in those groups would inherit the roles.
+
+But for a client, these groups may have business meaning or reflect organizational structure within their company, they aren't just a mean to group users together with the same roles, but rather fully functional teams that users can be part of with different roles.
+
+So UserA would be a `manager` in `group:iam`, but UserB could be a `developer`in `group:iam`.
+
+Namespaced roles would be specified when created, and loosely assigned based on the namespace, so let's say we have the following roles:
+
+`role_v1:/realm1/tenant1//iam/manager`
+`role_v1:/realm1/tenant1//iam/developer`
+`role_v1:/realm1/tenant1//devops/developer`
+`role_v1:/realm1/tenant1//devops/devops_role`
+
+Now these roles are assigned directly to `UserA` and `UserB` so that:
+
+- `UserA` has the `role_v1:/realm1/tenant1//iam/manager` and the `role_v1:/realm1/tenant1//devops/developer` roles.
+- `UserB` has the `role_v1:/realm1/tenant1//iam/developer` and the `role_v1:/realm1/tenant1//devops/developer` roles.
+
+Also, we should be able to still support giving everyone in a group an inherited role, such as:
+
+- `Group:devops` has the `role_v1:/realm1/tenant1//devops/devops_role` role.
+
+#### Representation in the tokens
+
+So now that we have a namespaced role attached to a user, the representation becomes inherently hierarchical, a **possible representation** in the tokens could look like:
+
+- For UserA
+
+```json
+{
+    ....
+    "roles": {
+        "tenants": {
+            "tenant1": {
+                "groups": {
+                    "iam": ["manager"],
+                    "devops": ["developer", "devops_role"]
+                }
+            }
+        }
+    }
+    ....
+}
+
+```
+
+- For UserB
+
+```json
+{
+    ....
+    "roles": {
+        "tenants": {
+            "tenant1": {
+                "groups": {
+                    "iam": ["developer"],
+                    "devops": ["developer", "devops_role"]
+                }
+            }
+        }
+    }
+    ....
+}
+```
+
+This also provides a way to isolate certain roles to specific tenants.
+
+- `role_v1:/realm1/tenant1//iam/developer` would only be visible in `tenant1` while
+- `role_v1:/realm1/tenant2//iam/somethingelse` would only be visible in `tenant2`.
+
+Or just to specific groups within the realm:
+
+- `role_v1:/realm1///iam/developer` would only be assignable to users in Group `iam`.
+- `role_v1:/realm1///devops/devops` would only be assignable to users in Group `devops`.
+
+### Use Case 2 - Custom, resource server specific roles
```

Comment:
Would be good to touch on how role namespaces align with client roles. In my mind client roles should be migrated to role namespaces, where a dedicated role namespace can easily be created for a specific client, or for a group of clients. With also making it more clear that you don't need to create a client just to represent some roles for a resource server.

Reply:
Good point. I think the biggest struggle in this whole namespaces spec is how to fit clients in.

We want to create a namespace that can be seen as a unique path to a role, but clients are not really part of that hierarchy necessarily.

Reply:
What I had in mind is that we'd reserve 'kc' as a special internal namespace, and for a dedicated/automatic namespace for a client it would just be `{role}:/kc/clients/{client-name}`